File: ruby/6kyu/sort_the_odd.rb
# You have an array of numbers.
# Your task is to sort ascending odd numbers but even numbers must be on their places.

# Zero isn't an odd number and you don't need to move it. If you have an empty array, you need to return it.

def sort_array(source_array)
  if source_array == []
    source_array
  else
    odds = source_array.select { |num| num.odd? }.sort

    result = []
    source_array.each do |number|
      if number.odd?
        result << odds[0]
        odds.shift
      else
        result << number
      end
    end
    result
  end
end

# p sort_array([5, 3, 2, 8, 1, 4, 0])
# [1, 3, 2, 8, 5, 4, 0]

# p sort_array([5, 3, 2, 8, 1, 4])
# [1, 3, 2, 8, 5, 4]

# p sort_array([])
# []

File: javascript/6kyu/sortTheOdd.js
/*
You will be given an array of numbers. You have to
sort the odd numbers in ascending order while
leaving the even numbers at their original positions.
*/

function identifyOdds(numbersArray) {
  const oddNumbers = [];

  for (let index = 0; index < numbersArray.length; index++) {
    const currentNumber = numbersArray[index];
    const oddNumber = currentNumber % 2 !== 0;

    if (oddNumber) {
      oddNumbers.push(currentNumber);
      numbersArray[index] = 'odd';
    }    
  }

  return [oddNumbers, numbersArray];
}

function insertodds(oddNumbers, array) {
  let oddNumbersIndex = 0;
  oddNumbers.sort((a, b) => a - b);

  return array.reduce((result, current) => {
    if (current === 'odd') {
      result.push(oddNumbers[oddNumbersIndex]);
      oddNumbersIndex++;
    } else {
      result.push(current);
    }

    return result;
  }, [])
}

function sortArray(numbersArray) {
  if (!numbersArray.length) return numbersArray;

  const [oddNumbers, markedArray] = identifyOdds(numbersArray);
  const sortedArray = insertodds(oddNumbers, markedArray);

  return sortedArray;
}

console.log(sortArray([5, 3, 2, 8, 1, 4]));
// [1, 3, 2, 8, 5, 4]

console.log(sortArray([5, 3, 1, 8, 0]));
// [1, 3, 5, 8, 0]

console.log(sortArray([]));
// []
